File: python/lru_cache.py
class Node:

    def __init__(self, key, val):
        self.key, self.val = key, val
        self.next, self.prev = None, None

class LRUCache:

    def __init__(self, capacity: int):
        self.cap = capacity
        self.cache = {}
        self.left, self.right = Node(0, 0), Node(0, 0)
        self.left.next, self.right.prev = self.right, self.left

    def get(self, key: int) -> int:
        if key in self.cache:
            self.evict(self.cache[key])
            self.insert(self.cache[key])
            return self.cache[key].val
        return -1
        

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            self.evict(self.cache[key])
        self.cache[key] = Node(key=key, val=value)
        self.insert(self.cache[key])

        if len(self.cache) > self.cap:
            lru = self.left.next
            self.evict(lru)
            del self.cache[lru.key]

    def insert(self, node: Node):
        node.next, node.prev = self.right, self.right.prev
        node.prev.next, node.next.prev = node, node

    def evict(self, node: Node):
        node.prev.next, node.next.prev = node.next, node.prev
        node.next, node.prev = None, None

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: C++/lru_cache.cpp
class Node {
public:
    int key;
    int val;
    Node* next;
    Node* prev;

    Node(int key, int val){
        this->key = key;
        this->val = val;
        this->next = nullptr;
        this->prev = nullptr;
    }
};

class LRUCache {

public:
    int cap;
    unordered_map<int, Node*> cache;
    Node* left;
    Node* right;

    LRUCache(int capacity) {
        this->cap = capacity;
        this->left = new Node(0, 0); 
        this->right = new Node(0, 0);
        this->left->next = this->right;
        this->right->prev = this->left;
    }
    
    int get(int key) {
        if (this->cache.contains(key)){
            evict(this->cache[key]);
            insert(this->cache[key]);
            return this->cache[key]->val;
        }
        return -1;
    }
    
    void put(int key, int value) {
        if (this->cache.contains(key)){
            evict(this->cache[key]);
        }
        this->cache[key] = new Node(key, value);
        insert(this->cache[key]);

        if (this->cache.size() > this->cap){
            Node* lru = this->left->next;
            evict(lru);
            this->cache.erase(lru->key);
        }
    }

    void insert(Node* node){
        node->next = this->right;
        node->prev = this->right->prev;

        node->next->prev = node;
        node->prev->next = node;
    }

    void evict(Node* node){
        node->prev->next = node->next;
        node->next->prev = node->prev;

        node->next = nullptr;
        node->prev = nullptr;
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */
